ERROR reading clientDir with macOS MS Remote Desktop.app with devel

### xrdp version

0.9.80

### Detailed xrdp version, build options
```
  Configure options:
      --localstatedir=/var
      --enable-strict-locations
      --enable-pam-config=freebsd
      --enable-jpeg
      --enable-pixman
      --enable-rfxcodec
      --enable-painter
      --enable-imlib2
      --prefix=/opt/xrdp
      --enable-fuse
      --enable-devel-logging
      --disable-silent-rules
      CC=clang16
      CFLAGS= -I/usr/local/include
      LDFLAGS= -L/usr/local/lib -L/usr/lib
      OPENSSL_CFLAGS=-I/usr/include
      OPENSSL_LIBS=-lssl -lcrypto

  Compiled with OpenSSL 1.1.1t-freebsd  7 Feb 2023
```

### Operating system & version

FreeBSD 13.2

### Installation method

git clone & make install

### Which backend do you use?

xorgxrdp

### What desktop environment do you use?

fvwm3

### Environment xrdp running on

FreeBSD 13.2 jail

### What's your client?

macOS Microsoft Remote Desktop

### Area(s) with issue?

_No response_

### Steps to reproduce

RDP to the server with macOS MS Remote Desktop.app (Version 10.9.4 (2161))

### ✔️ Expected Behavior

To be presented with the RDP login screen.

### ❌ Actual Behavior

With macOS MS RDP application and devel 50cff2eb fails to present the login screen with the following error in xrdp.log:

```
[2023-11-11T23:19:09.411+0000] [DEBUG] [xrdp_sec_process_logon_info(xrdp_sec.c:1087)] Client supplied directory:
[2023-11-11T23:19:09.423+0000] [TRACE] [ts_info_utf16_in(xrdp_sec.c:799)] ts_info_utf16_in: uni_len 26, dst_len 256
[2023-11-11T23:19:09.442+0000] [TRACE] [ts_info_utf16_in(xrdp_sec.c:799)] ts_info_utf16_in: uni_len -2, dst_len 256
[2023-11-11T23:19:09.456+0000] [ERROR] [ts_info_utf16_in(xrdp_sec.c:821)] ts_info_utf16_in: bad terminator. Expected 0, got 300
[2023-11-11T23:19:09.470+0000] [ERROR] [xrdp_sec_process_logon_info(xrdp_sec.c:1113)] ERROR reading clientDir
```

It appears if ts_info_utf16_in is called when src_bytes is 0 uni_len is -2... this didn't seem correct.  Added a check to not perform term check if src_bytes is <= 0.  This appears to allow the macOS MS RDP client to login however I'm not sure if this is a client bug or possible issue with ts_info_utf16_in or previous processing of ip data from the stream.

### Anything else?

_No response_

--- a/common/ms-rdpbcgr.h
+++ b/common/ms-rdpbcgr.h
@@ -131,6 +131,9 @@
 #define RDP_LOGON_LEAVE_AUDIO          0x2000
 #define RDP_LOGON_RAIL                 0x8000
 
+/* Extended Info Packet: clientAddress (2.2.1.11.1.1.1) */
+#define EXTENDED_INFO_MAX_CLIENT_ADDR_LENGTH 80
+
 /* Extended Info Packet: performanceFlags (2.2.1.11.1.1.1) */
 /* TODO: to be renamed */
 #define RDP5_DISABLE_NOTHING           0x00

--- a/libxrdp/xrdp_sec.c
+++ b/libxrdp/xrdp_sec.c
@@ -833,18 +833,15 @@ static int
 xrdp_sec_process_logon_info(struct xrdp_sec *self, struct stream *s)
 {
     int flags = 0;
-    int len_domain = 0;
-    int len_user = 0;
-    int len_password = 0;
-    int len_program = 0;
-    int len_directory = 0;
-    int len_ip = 0;
-    int len_dll = 0;
-    char tmpdata[256];
+    unsigned int len_domain = 0;
+    unsigned int len_user = 0;
+    unsigned int len_password = 0;
+    unsigned int len_program = 0;
+    unsigned int len_directory = 0;
+    unsigned int len_clnt_addr = 0;
+    unsigned int len_clnt_dir = 0;
     const char *sep;
 
-    /* initialize (zero out) local variables */
-    g_memset(tmpdata, 0, sizeof(char) * 256);
     if (!s_check_rem_and_log(s, 8, "Parsing [MS-RDPBCGR] TS_INFO_PACKET"))
     {
         return 1;
@@ -1097,22 +1094,33 @@ xrdp_sec_process_logon_info(struct xrdp_sec *self, struct stream *s)
         }
         /* TS_EXTENDED_INFO_PACKET required fields */
         in_uint8s(s, 2);         /* clientAddressFamily */
-        in_uint16_le(s, len_ip);
-        if (ts_info_utf16_in(s, len_ip - 2, tmpdata, sizeof(tmpdata)) != 0)
+        in_uint16_le(s, len_clnt_addr);
+        if (len_clnt_addr > EXTENDED_INFO_MAX_CLIENT_ADDR_LENGTH ||
+                !s_check_rem(s, len_clnt_addr))
         {
-            LOG(LOG_LEVEL_ERROR, "ERROR reading ip");
+            LOG(LOG_LEVEL_ERROR, "clientAddress is too long (%u bytes)",
+                len_clnt_addr);
             return 1;
         }
+        // The clientAddress is currently unused. [MS-RDPBCGR] requires
+        // a mandatory null terminator, but some clients set
+        // len_clnt_addr == 0 if this field is missing. Allow for this
+        // in any future implementation.
+        in_uint8s(s, len_clnt_addr); // Skip Unicode clientAddress
+
         if (!s_check_rem_and_log(s, 2, "Parsing [MS-RDPBCGR] TS_EXTENDED_INFO_PACKET clientDir"))
         {
             return 1;
         }
-        in_uint16_le(s, len_dll);
-        if (ts_info_utf16_in(s, len_dll - 2, tmpdata, sizeof(tmpdata)) != 0)
+        in_uint16_le(s, len_clnt_dir);
+        if (len_clnt_dir > INFO_CLIENT_MAX_CB_LEN ||
+                !s_check_rem(s, len_clnt_dir))
         {
-            LOG(LOG_LEVEL_ERROR, "ERROR reading clientDir");
+            LOG(LOG_LEVEL_ERROR, "clientDir is too long (%u bytes)", len_clnt_dir);
             return 1;
         }
+        in_uint8s(s, len_clnt_dir); // Skip Unicode clientDir
+
         LOG_DEVEL(LOG_LEVEL_TRACE, "Received [MS-RDPBCGR] TS_EXTENDED_INFO_PACKET "
                   "<Required Fields> clientAddressFamily (ignored), "
                   "cbClientAddress (ignored), clientAddress (ignored), "

